libgnucash/engine/gnc-commodity.cpp

```diff
@@ -488,15 +464,38 @@ gnc_quote_source_get_type (const gnc_quote_source *source)
 gint
 gnc_quote_source_get_index (const gnc_quote_source *source)
 {
-    ENTER("%p", source);
     if (!source)
     {
-        LEAVE("bad source");
+        PWARN ("bad source");
         return 0;
     }
 
-    LEAVE("index is %ld", source->get_index());
-    return source->get_index();
+    switch (source->get_type())
+    {
+    case SOURCE_CURRENCY:
+        return 0;
+    case SOURCE_SINGLE:
+        for (auto i = 0; i < num_single_quote_sources; ++i)
+            if (&single_quote_sources[i] == source)
+                return i;
+        break;
```

Comment:
I think this can work. If you like yet another step further in the C++ conversion, you may want to look into converting the static arrays of quote sources into a std::vector instead. That would open the doors for some  C++ algorithm fun. With a vector you could replace the generic for-loops with something more expressive like:
```
    auto quote_it = std::find (single_quote_sources.begin(), single_quote_sources.end(), source);
    auto i = std::distance (single_quote_sources.begin(), quote_it);
    if (i == single_quote_sources.size())
        i = 0; // no match found
    return i;
```
Perhaps the std::find will have to be std::find_if with a custom search lambda, but I think it may work as suggested. With a little more refactoring you can reduce the switch to only select the vector to work on as the std::vectors would all be the same type.

```
    auto sources_vec = single_quote_sources(); // can be whatever, just to initialize the vector
    switch (source->get_type())
    {
    case SOURCE_SINGLE:
        sources_vec = single_quote_sources;
        break;
    case SOURCE_MULTI:
        ....
    }
  
    auto quote_it = std::find (sources_vec.begin(), sources_vec.end(), source);
    auto i = std::distance (sources_vec.begin(), quote_it);
    if (i == sources_vec.size())
        i = 0; // no match found
    return i;
```
  
And if it were me, I'd even make the currency_source a std::vector instead of just a struct. That would allow you to treat it just the same as the other vectors everywhere.

With vectors there will be plenty of other generic for loops you can replace with more expressive algorithms in gnc-commodity.cpp by the way. Just tell me when to stop :wink:

Reply:
Changing the quote sources to std::vector will also help #1626. A good next step (not for this PR) will move the F::Q specific code out of gnc-commodity to app-utils/gnc-quotes.cpp or maybe a new app-utils/gnc-finance-quote.cpp.

Reply:
I had forgotten about that PR...

While glancing over it I noticed a commit that does away with the `old_internal_name` variable. That's a nice cleanup standing on its own. Perhaps we could cherry-pick that commit to maint and rebase this PR on top of it ? That honours the authors work and avoids having to redo it after this PR (I anticipate only a few easy to resolve merge conflicts in this rebase, the other parts of #1626 touching gnc-commodity.cpp will need more significant rework).

Reply:
I couldn't resist and tried this cherry-pick and merge myself. Turns out it affects almost every commit in this PR... Good news is all merge conflicts are trivial. If you want I can force-push this work to this PR, so you don't have to redo the merges.

Reply:
> I couldn't resist and tried this cherry-pick and merge myself. Turns out it affects almost every commit in this PR... Good news is all merge conflicts are trivial. If you want I can force-push this work to this PR, so you don't have to redo the merges.

Go ahead :) I'm not entirely sure why old_internal_name exists anyway.

Reply:
Ok, done. I have also added two commits of my own. One with just minor NULL/nullptr cleanups and the other with two more c++ algorithm conversions.

Reply:
> While glancing over it I noticed a commit that does away with the `old_internal_name` variable. That's a nice cleanup standing on its own. Perhaps we could cherry-pick that commit to stable and rebase this PR on top of it ?

@vincentl this is now done. Your `old_internal_name` cleanup merged in as f5b7bf3401, and this branch is rebased on stable. I've added a tiny modification to modify test to illustrate the change.